File: 1035-cousins-in-binary-tree/cousins-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isCousins(self, root: Optional[TreeNode], x: int, y: int) -> bool:
        self.xp = self.yp = None
        self.xd = self.yd = -1

        def dfs(node, parent, depth):
            if not node:
                return
            
            if node.val == x:
                self.xp = parent
                self.xd = depth
            if node.val == y:
                self.yp = parent
                self.yd = depth

            dfs(node.left, node, depth + 1)
            dfs(node.right, node, depth + 1)

        dfs(root, None, 0)

        return self.xd == self.yd and self.xp != self.yp

File: 1035-cousins-in-binary-tree/cousins-in-binary-tree.js
var isCousins = function (root, x, y) {
    let xp, yp
    let xd, yd

    function dfs(node, p, depth) {
        if (!node) return;

        if (node.val === x) {
            xp = p
            xd = depth
        }
        if (node.val === y) {
            yp = p
            yd = depth
        }

        dfs(node.left, node, depth+1);
        dfs(node.right, node, depth+1);
    }

    dfs(root, null, null)

    return (xp !== yp) && (xd === yd)
};

// var isCousins3333 = function (root, x, y) {
//     let op = {};
//     var rec = function (r, d) {
//         if (r.left) {
//             op[r.left.val] = { p: r.val, d };
//             rec(r.left, d + 1);
//         }
//         if (r.right) {
//             op[r.right.val] = { p: r.val, d };
//             rec(r.right, d + 1);
//         }
//     };

//     rec(root, 0);
//     return ((op[x]?.p !== op[y]?.p) && (op[x]?.d === op[y]?.d))
// };
